### PR TITLE
Regridding performance

### LINKED ISSUE #675: Regridding is very slow for large grids
Regridding involving large grids (e.g. 4200 x 43200) is very slow  - sometimes taking hours.

Much of this problem is the time taken to calculate the weights (which is done non-laziliy), and to update the source and destination masks based on the weights' values.

This will be remedied in an upcoming PR.

### PR DESCRIPTION
**Pull request by davidhassell**
Fixes #675

### CODE REVIEW

File: cf/regrid/regrid.py [hunk 1]
```diff
@@ -1453,92 +1480,178 @@ def create_esmpy_weights(
...
-            element.
+            the instance is made available as the list's last element.
+
+        weights_file: `str` or `None`, optional
+            Provide a netCDF file that contains, or will contain, the
+            regridding weights. If `None` (the default) then the
+            weights are computed in memory for regridding between the
+            source and destination grids, and no file is created.
+
+            If set to a file path that does not exist then thea
```
Comment (sadielbartholomew) [inline cf/regrid/regrid.py:1499]:
```suggestion
            If set to a file path that does not exist then the
```

File: cf/regrid/regridoperator.py [hunk 2]
```diff
@@ -324,13 +333,28 @@ def start_index(self):
 
     @property
     def weights(self):
-        """The 1-d array of the regridding weights.
+        """The 1-d array of the regridding weights, or `None`
```
Comment (sadielbartholomew) [inline cf/regrid/regridoperator.py:336]:
```suggestion
        """The 1-d array of the regridding weights, or `None`.
```

**Review by sadielbartholomew — APPROVED** [review]
Excellent. Logically this should speed up the regridding, so assuming the practical performance tests that Robin etc. are conducting don't indicate otherwise, I am happy for this to be merged (once minor comments are addressed).

File: cf/test/test_regrid.py [hunk 3]
```diff
@@ -9,6 +10,23 @@
...
+
+def _remove_tmpfiles():
+    """Try to remove defined temporary files by deleting their paths."""
+    for f in tmpfiles:
+        try:
+            os.remove(f)
+        except OSError:
+            pass
+
+
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/test/test_regrid.py:29]: Quite right

File: cf/test/test_regrid.py [hunk 4]
```diff
@@ -726,6 +745,34 @@ def test_Field_regrid_chunks(self):
...
+    @unittest.skipUnless(esmpy_imported, "Requires esmpy/ESMF package.")
+    def test_Field_regrid_weights_file(self):
+        """Regridding creation/use of weights file"""
+        dst = self.dst
+        src = self.src
+        tmpfile = "delme.nc"
+        try:
+            os.remove(tmpfile)
+        except OSError:
+            pass
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/test/test_regrid.py:760]: The `try .. except ...` will also remove the file if it does exist (for instance, by being created by a previous test). The line `tmpfile = "delme.nc"` is a red herring that was left over from testing, and I have removed it!

[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/test/test_regrid.py:761]: Ah, I see. Thanks for clarifying.

**Pull request merged**